stacker/util.py

```diff
@@ -547,53 +603,80 @@ def fetch_s3_package(self, config):
+        for suffix, klass in extractor_map.iteritems():
+            if config['key'].endswith(suffix):
+                extractor = klass()
+                dir_name = self.sanitize_uri_path(
+                    "s3-%s-%s" % (config['bucket'],
+                                  config['key'][:-len(suffix)])
+                )
+                break
+
+        if extractor is None:
             raise ValueError(
                 "Archive type could not be determined for S3 object \"%s\" "
                 "in bucket %s." % (config['key'], config['bucket'])
             )
 
+        session = get_session(region=None)
+        extra_s3_args = {}
+        if config.get('requester_pays', False):
+            extra_s3_args['RequestPayer'] = 'requester'
+
         # We can skip downloading the archive if it's already been cached
+        if config.get('use_latest', True):
+            try:
+                modified_date = session.client('s3').head_object(
+                    Bucket=config['bucket'],
+                    Key=config['key'],
+                    **extra_s3_args
+                )['LastModified']
+            except botocore.exceptions.ClientError as client_error:
+                logger.error("Error checking modified date of "
+                             "s3://%s/%s : %s",
+                             config['bucket'],
+                             config['key'],
+                             client_error)
+                sys.exit(1)
+            dir_name += "-%s" % modified_date.strftime(self.CACHE_DATE_FORMAT)
         cached_dir_path = os.path.join(self.package_cache_dir, dir_name)
         if not os.path.isdir(cached_dir_path):
+            logger.debug("Remote package s3://%s/%s does not appear to have "
+                         "been previously downloaded - starting download and "
+                         "extraction to %s",
+                         config['bucket'],
+                         config['key'],
+                         cached_dir_path)
             tmp_dir = tempfile.mkdtemp(prefix='stacker')
             tmp_package_path = os.path.join(tmp_dir, dir_name)
             try:
-                tmp_archive_path = os.path.join(tmp_dir,
-                                                dir_name + archive_type)
-                s3_resource = get_session(region=None).resource('s3')
-                if config.get('requester_pays', False):
-                    extra_args = {'RequestPayer': 'requester'}
-                else:
-                    extra_args = None
-                s3_resource.Bucket(config['bucket']).download_file(
+                extractor.archive = os.path.join(
+                    tmp_dir,
+                    dir_name + extractor.extension()
+                )
+                session.resource('s3').Bucket(config['bucket']).download_file(
                     config['key'],
-                    tmp_archive_path,
-                    ExtraArgs=extra_args
+                    extractor.archive,
```

Comment:
Just to future-proof this some, can we have a method called `set_archive` on the base `Extractor` class that does this?  That way we can override it later if we need to in an easier way.